### PR TITLE
Multimodal Contrastive loss

### PR DESCRIPTION
**Pull request by lucidtronix**
Adds a contrastive Loss layer that contrasts modalities which have been paired via the `--pairs` argument, building on to the existing PairedLoss merge Block.  This is in contrast to ;) contrastive loss across time as in PCLR...

### CODE REVIEW

File: ml4h/models/merge_blocks.py [hunk 1]
```diff
@@ -119,13 +122,16 @@ def __init__(
...
+            batch_size: int = 4,
             **kwargs,
     ):
         self.pairs = pairs
         if pair_loss == 'cosine':
             self.loss_layer = CosineLossLayer(pair_loss_weight)
         elif pair_loss == 'euclid':
             self.loss_layer = L2LossLayer(pair_loss_weight)
+        elif pair_loss == 'contrastive':
+            self.loss_layer = ContrastiveLossLayer(pair_loss_weight, batch_size)
```
Comment (ndiamant) [inline ml4h/models/merge_blocks.py:135]: maybe include `else: raise ValueError(f"{pair_loss} is not a valid loss type")` after this incase of mistyped loss

File: ml4h/models/merge_blocks.py [hunk 2]
```diff
@@ -153,6 +159,22 @@ def pairwise_cosine_difference(t1, t2):
     return K.mean(tf.acos(dot))
 
 
+def contrastive_difference(left: Tensor, right: Tensor, batch_size: int, temperature: Tensor):
```
Comment (ndiamant) [inline ml4h/models/merge_blocks.py:162]: am i remembering right that you were getting some numerical issues in this loss? Maybe the trick of subtracting a large constant before applying the softmax could help

**Review by ndiamant — APPROVED** [review]
Small nit and numerical suggestion. Once addressed or pondered and rejected, no need to come back for feedback again

File: ml4h/models/merge_blocks.py (hunk 1 above)
Comment (lucidtronix) [reply ml4h/models/merge_blocks.py:137]: Done!

File: ml4h/models/merge_blocks.py (hunk 2 above)
Comment (lucidtronix) [reply ml4h/models/merge_blocks.py:164]: cool Idk that trick :)

**Pull request merged**